### PR TITLE
Implement CUDA_KERNEL_ASSERT for MSVC

### PR DESCRIPTION
**Pull request by peterjc123**
Tested locally on CPU/GPU + Debug/Release.

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
@peterjc123 awesome! Maybe with asserts enabled on windows this test https://github.com/pytorch/pytorch/blob/0e8c65f756c5c24620c843d70f4ceca6a5faeb24/test/test_cuda.py#L1726 can be reenabled?

**Comment by peterjc123** [comment]
I'll try to re-enable it.

**Comment by peterjc123** [comment]
cc @gchanan Maybe we should get this into v1.5.1.

File: c10/macros/Macros.h [hunk 1]
```diff
@@ -216,9 +216,22 @@ constexpr uint32_t CUDA_THREADS_PER_BLOCK_FALLBACK = 256;
...
+#endif // __CUDA_ARCH__
+ void _wassert(
+    wchar_t const* _Message,
+    wchar_t const* _File,
+    unsigned _Line);
+}
+#endif
+#define CUDA_KERNEL_ASSERT(cond)                                                                                 \
+  if (C10_UNLIKELY(!(cond))) {                                                                                   \
+    (void)((!!(cond)) || (_wassert(_CRT_WIDE(#cond), _CRT_WIDE(__FILE__), static_cast<unsigned>(__LINE__)), 0)); \
```
Comment (malfet) [inline c10/macros/Macros.h:233]:
```suggestion
    (void)(_wassert(_CRT_WIDE(#cond), _CRT_WIDE(__FILE__), static_cast<unsigned>(__LINE__)), 0)); \
No need to check cond again, as on line 232 it was already established, that `cond` is false
```suggestion
    (void)((!!(cond)) || (_wassert(_CRT_WIDE(#cond), _CRT_WIDE(__FILE__), static_cast<unsigned>(__LINE__)), 0)); \
```

**Review by malfet — APPROVED** [review]

File: c10/macros/Macros.h (hunk 1 above)
Comment (peterjc123) [reply c10/macros/Macros.h:233]: Nice catch.

Comment (peterjc123) [reply c10/macros/Macros.h]: @malfet Done!

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**